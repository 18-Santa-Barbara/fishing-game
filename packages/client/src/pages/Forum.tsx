import {
  Table,
  Box,
  TableContainer,
  Typography,
  TableRow,
  TableCell,
  TableHead,
  TableBody,
  Paper,
  Button,
  DialogActions,
  Dialog,
  DialogContent,
  DialogTitle,
  TextField,
} from '@mui/material';

import AddIcon from '@mui/icons-material/Add';
import { makeStyles } from '@mui/styles';
import { useState } from 'react';
import { useGetForumQuery, useSetForumMutation } from '../services/forumApi';
import { useGetUserQuery } from '../services/userApi';
import { useNavigate } from 'react-router-dom';

export type PostType = {
  id?: number;
  title: string;
  author: string;
  updateTime: string;
  body: string;
};

const useStyles = makeStyles(() => ({
  root: { marginTop: '30px' },
  paper: {
    boxShadow: '0px 0px 6px rgb(0 0 0 / 14%)',
    borderRadius: '12px',
    position: 'absolute',
    top: '50%',
    left: '50%',
    minWidth: '750px',
    transform: 'translate(-50%, -50%)',
    padding: '16px',
  },
  btn: {
    margin: '12px 0',
    width: '33%',
  },
  centerText: {
    textAlign: 'center',
  },
  table: {},
}));

function Forum() {
  const { data: user } = useGetUserQuery();
  const { data: forum, isLoading: isLoadingForum } = useGetForumQuery();
  const [addForumPost, response] = useSetForumMutation();
  const [openNewPost, setOpenNewPost] = useState(false);
  const [title, setTitle] = useState('');
  const [body, setBody] = useState('');
  const classes = useStyles();
  const navigate = useNavigate();

  const handleClickOpenNewPost = () => {
    setOpenNewPost(true);
  };

  const handleClickCloseNewPost = () => {
    setOpenNewPost(false);
  };

  const getTime = (date: string | number) => {
    return `${date}`;
  };

  const navigateToComments = (id: any) => {
    navigate(`/comments/${id}`);
  };

  const handleClickNewPost = () => {
    const forumData: PostType = {
      title: title,
      author: user.login,
      updateTime: new Date().toDateString(),
      body: body
    };

    addForumPost(forumData)
      .then(() => {
        setOpenNewPost(false);
      })
      .catch(err => console.warn('error: ', err));
  };

  if (isLoadingForum) {
    return <div>Loading</div>;
  }

  return (
    <Box className={classes.paper}>
      <Typography variant="h3" className={classes.centerText}>
        Forum
      </Typography>
      <Dialog
        open={openNewPost}
        onClose={handleClickCloseNewPost}
        aria-labelledby="form-dialog-title">
        <DialogTitle id="form-dialog-title">New post</DialogTitle>
        <DialogContent>
          <TextField
            autoFocus
            margin="dense"
            id="title"
            label="Title"
            type="text"
            fullWidth
            onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
              setTitle(e.target.value)
            }
          />
          <TextField
            autoFocus
            margin="dense"
            id="body"
            label="Body"
            type="text"
            multiline
            fullWidth
            onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
              setBody(e.target.value)
            }
          />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClickCloseNewPost} color="primary">
            Cancel
          </Button>
          <Button onClick={handleClickNewPost} color="primary">
            Create
          </Button>
        </DialogActions>
      </Dialog>
      <TableContainer component={Paper}>
        <Table
          className={classes.table}
          size="small"
          aria-label="a dense table">
          <TableHead>
            <TableRow>
              <TableCell>Title</TableCell>
              <TableCell>Author</TableCell>
              <TableCell>Update date</TableCell>
              <TableCell>
                <Button
                  variant="contained"
                  color="primary"
                  className={classes.btn}
                  endIcon={<AddIcon />}
                  onClick={handleClickOpenNewPost}
                  fullWidth>
                  New
                </Button>
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {forum &&
              !!forum.length &&
              forum.map(row => (
                <TableRow key={row.title}>
                  <TableCell component="th" scope="row">
                    {row.title}
                  </TableCell>
                  <TableCell>{row.author}</TableCell>
                  <TableCell>{getTime(row.updateTime)}</TableCell>
                  <TableCell>
                    <Button
                      color="secondary"
                      onClick={() => {
                        navigateToComments(row.id);
                      }}>
                      read
                    </Button>
                  </TableCell>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </TableContainer>
    </Box>
  );
}

export default Forum;
