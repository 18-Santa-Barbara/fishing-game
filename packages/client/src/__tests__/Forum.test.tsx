import * as renderer from 'react-test-renderer';

jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

import Forum from '../pages/Forum';

it('Forum page', () => {
  const tree = renderer.create(<Forum />).toJSON();
  expect(tree).toMatchSnapshot();
});
