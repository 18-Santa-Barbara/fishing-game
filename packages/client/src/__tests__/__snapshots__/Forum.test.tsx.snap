// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forum page 1`] = `
<div
  className="makeStyles-paper-2 MuiBox-root css-0"
>
  <h3
    className="MuiTypography-root MuiTypography-h3 makeStyles-centerText-4 css-gepadz-MuiTypography-root"
  >
    Forum
  </h3>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
  >
    <table
      aria-label="a dense table"
      className="MuiTable-root makeStyles-table-5 css-rqglhn-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
            scope="col"
          >
            Title
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
            scope="col"
          >
            Author
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
            scope="col"
          >
            Update time
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
            scope="col"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth makeStyles-btn-3 css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              New
              <span
                className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
              </span>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            scope="row"
          >
            Title 1
          </th>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Test user
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            01.01.2020, 03:00:00
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium css-1d3f8j8-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              read
            </button>
          </td>
        </tr>
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            scope="row"
          >
            Title 2
          </th>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Test user 2
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            01.01.2020, 03:00:00
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium css-1d3f8j8-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              read
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
